forge/test/mlir/llama/test_llama_inference.py

```diff
@@ -120,3 +121,346 @@ def test_llama_inference_cache_cpu():
+    inputs = tokenizer(
+        prompt,
+        return_tensors="pt",
+        max_length=max_sequence_length,
+        pad_to_max_length=True,
+        truncation=True,
+    )
+    input_ids = inputs.input_ids
+    attention_mask = inputs.attention_mask
+
+    non_padding_seq_len = int(torch.sum(attention_mask))
+    padding_seq_len = attention_mask.shape[1] - non_padding_seq_len
+
+    # Run Prefill on CPU with no cache to get the initial logits
+    logits = framework_model(input_ids=input_ids, attention_mask=attention_mask)
+
+    # Take the last non padding token index in logits for the next token
+    next_token_logits = logits[0][:, non_padding_seq_len - 1, :]
+    next_token = torch.argmax(next_token_logits, dim=-1)
+
+    # Update the input_ids with predicted token from prefill in the first padding token index
+    input_ids[:, non_padding_seq_len] = next_token
+    attention_mask[:, non_padding_seq_len] = 1
+
+    non_padding_seq_len = int(torch.sum(attention_mask))
+    padding_seq_len = attention_mask.shape[1] - non_padding_seq_len
+
+    # Compile the model on TT
+    compiled_model = forge.compile(
+        framework_model, sample_inputs=[input_ids, attention_mask]
+    )
+
+    # Run decode stage on TT device and generate tokens by appending predicted token into sequence of input tokens
+    # untill the a specified maximum number of new tokens is reached or an end-of-sequence token is encountered.
+    max_new_tokens = padding_seq_len
+    for idx in range(max_new_tokens):
+
+        model_inputs = [input_ids, attention_mask]
+
+        # CPU Inference
+        framework_output = framework_model(
+            input_ids=input_ids, attention_mask=attention_mask
+        )
+
+        # Run on TT device
+        tt_output = compiled_model(*model_inputs)
+        tt_output = [tt_out.to("cpu") for tt_out in tt_output]
+
+        # Validate TT result with Framework
+        assert all(
+            [
+                compare_with_golden_pcc(golden=fw_out, calculated=tt_out, pcc=0.99)
+                for fw_out, tt_out in zip(framework_output, tt_output)
+            ]
+        )
+
+        next_token_index = non_padding_seq_len + idx
+        next_token_logits = tt_output[0][:, next_token_index - 1, :]
+        next_token = torch.argmax(next_token_logits, dim=-1)
+
+        if next_token == tokenizer.eos_token_id:
+            break
+
+        input_ids[:, next_token_index] = next_token
+        attention_mask[:, next_token_index] = 1
+
+    # Generated text
+    generated_text = tokenizer.decode(input_ids[0], skip_special_tokens=True)
+    print("generated_text=", generated_text)
+
+
+@pytest.mark.xfail()
+def test_llama_prefill_on_cpu_decode_on_tt_cache():
+
+    # Load Llama 3B model and tokenizer
+    model_path = "openlm-research/open_llama_3b"
+    framework_model, tokenizer = load_model_and_tokenizer(
+        model_path=model_path, use_cache=True, return_dict=True
+    )
+
+    class LlamaModelWrapper(torch.nn.Module):
```

Comment:
Can we move this class deffinition out of the test...

Reply:
Sure @dgolubovicTT